src/block/block_cache.c

```diff
@@ -442,90 +447,91 @@ __wt_blkcache_put(WT_SESSION_IMPL *session, wt_off_t offset, size_t size, uint32
      */
     if ((blkcache->cache_on_checkpoint == false) && checkpoint_io) {
         WT_STAT_CONN_INCR(session, block_cache_bypass_chkpt);
-        return (WT_BLKCACHE_BYPASS);
+        return (0);
     }
 
     /* Bypass on high overhead */
     if (__blkcache_high_overhead(session) == true) {
         WT_STAT_CONN_INCR(session, block_cache_bypass_overhead_put);
-        return (WT_BLKCACHE_BYPASS);
+        return (0);
     }
-    /*
-     * Allocate space in the cache outside of the critical section. In the unlikely event that we
-     * fail to allocate metadata, or if the item exists and the caller did not check for that prior
-     * to calling this function, we will free the space.
-     */
-    WT_RET(__blkcache_alloc(session, size, &data_ptr));
 
     /*
-     * We will use the item id structure as a byte array for (1) computing a hash, and (2) for
-     * comparing with the items in the hash table.
+     * Allocate and initialize space in the cache outside of the critical section. In the unlikely
+     * event that we fail an allocation, free the space. NVRAM allocations can fail if there's no
+     * available memory, treat it as a cache-full failure.
      */
-    id.checksum = checksum;
-    id.offset = offset;
-    id.size = (uint32_t)size;
-    id.fid = S2BT(session)->id;
-    hash = __wt_hash_city64(&id, sizeof(id));
-
+    WT_RET(__blkcache_alloc(session, data->size, &data_ptr));
+    if (data_ptr == NULL)
+        return (0);
+    WT_ERR(__wt_calloc_one(session, &blkcache_store));
+    blkcache_store->fid = S2BT(session)->id;
+    WT_ASSERT(session, addr_size <= sizeof(blkcache_store->addr));
```

Comment:
I don't understand what this assert is doing.

Reply:
I created a fixed-size array to hold an address cookie in the `WT_BLKCACHE_ITEM` structure (`uint32_t addr[WT_INTPACK64_MAXSIZE * 3]`), and that assert guarantees we never see an address cookie larger than the array (the address cookie can theoretically be up to 255B, or `WT_BTREE_MAX_ADDR_COOKIE` in length, and that would be prohibitively large to pre-allocate in these structures, so I did something that's not very good and asserted I never got caught).

The real fix is to allocate exactly what we need to store the address cookie as part of the `WT_BLKCACHE_ITEM` structure itself, that means we can handle any size address cookie and still only do a single allocation.

Reply:
@fedorova, update, I pushed the flexible array member fix in a12c8620c.

Reply:
Ah, I see, I thought that addr was a pointer, not a fixed-sized. array.